src/OrchardCore.Modules/OrchardCore.AutoSetup/AutoSetupStartupFilter.cs

```diff
@@ -0,0 +1,88 @@
+using System;
+using System.ComponentModel.DataAnnotations;
+using System.Linq;
+using System.Text;
+using Microsoft.AspNetCore.Builder;
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using OrchardCore.AutoSetup.Options;
+using OrchardCore.Environment.Shell.Scope;
+
+namespace OrchardCore.AutoSetup
+{
+    /// <summary>
+    /// The filter which registers <see cref="AutoSetupMiddleware"/> to setup the site.
+    /// </summary>
+    public class AutoSetupStartupFilter : IStartupFilter
+    {
+        /// <summary>
+        /// The logger.
+        /// </summary>
+        private readonly ILogger _logger;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="AutoSetupStartupFilter"/> class.
+        /// </summary>
+        /// <param name="logger"> The logger. </param>
+        public AutoSetupStartupFilter(ILogger<AutoSetupStartupFilter> logger)
+        {
+            this._logger = logger;
+        }
+
+        /// <summary>
+        /// The configure a middleware pipeline
+        /// </summary>
+        /// <param name="next"> The next middleware in the execution pipeline. </param>
+        /// <returns> The <see cref="Action"/>. </returns>
+        public Action<IApplicationBuilder> Configure(Action<IApplicationBuilder> next)
+        {
+            var scopedServices = ShellScope.Services;
```

Comment:
I'm not sure, I saw it is being used in the other parts, @jtkech can you suggest, please?

Reply:
@rserj @deanmarcussen 

Good question ;)

Yes it is used in some places, but because it is less DI firendly we could say it is better to use the DI when it is possible, we use it e.g. for `ISiteService` that is a tenant singleton for some reasons but that needs to be executed in a tenant / shell scope context, this to resolve scoped services.

Here it's the same kind of thing, yes `IStartupFilter` are executed under a shell scope context while building the tenant pipeline, but they are resolved through the tenant / shell container, not through a shell scope, we could have done it but we opted to keep for a tenant the pattern of `IStartupFilter` as for a regular "app" (here the tenant as an app with its own shell container).

            // Create a nested pipeline to configure the tenant middleware pipeline
            var startupFilters = appBuilder.ApplicationServices.GetService<IEnumerable<IStartupFilter>>();


So here, if you inject `IServiceProvider` it would be related to the tenant / shell container, not the one of the current shell scope, so you would not be able to resolve scoped services. So yes, using `ShellScope.Services;` is a solution, the other solution is to inject `IHttpContextAccessor` to get the `RequestServices` that we take care to be the services of the shell scope if there is a current one ;)

Both are using an `AsyncLocal` object to be retrieved in a given async execution flow, the current `HttpContext` or current `ShellScope`. Personnaly i would use `ShellScope`, but `IHttpContextAccessor` is more aspnetcore friendly.

Hmm, maybe another solution through a module startup `Configure()` that passes a scoped service provider, and with a lower `ConfigureOrder` to get called earlier. Hmm, i could get an auto setup working by only using the config sources stack + a few lines of code https://github.com/OrchardCMS/OrchardCore/pull/4567#issuecomment-544059139, but i don't remember what was missing.

i would need to take a deeper look on your PR, i will do soon when i will have time.